lib/datadog/core/telemetry/collector.rb

```diff
@@ -0,0 +1,244 @@
+          'tracing.partial_flush.enabled'.freeze,
+          'tracing.partial_flush.min_spans_threshold'.freeze,
+          'tracing.priority_sampling'.freeze,
+          'tracing.report_hostname'.freeze,
+          'tracing.sampling.default_rate'.freeze,
+          'tracing.sampling.rate_limit'.freeze
+        ].freeze
+
+        def additional_payload_variables
+          # Whitelist of configuration options to send in additional payload object
+          config_options = Datadog.configuration.to_hash
+          config_options_to_keep = {}
+          TARGET_OPTIONS.each do |option|
+            config_options_to_keep[option] = format_configuration_value(config_options[option])
+          end
+
+          # Add some more custom additional payload values here
+          config_options_to_keep['tracing.auto_instrument.enabled'.freeze] = !defined?(Datadog::AutoInstrument::LOADED).nil?
+          config_options_to_keep['tracing.writer_options.buffer_size'.freeze] =
+            format_configuration_value(Datadog.configuration.tracing.writer_options[:buffer_size])
+          config_options_to_keep['tracing.writer_options.flush_interval'.freeze] =
+            format_configuration_value(Datadog.configuration.tracing.writer_options[:flush_interval])
+          config_options_to_keep['logger.instance'.freeze] = Datadog.configuration.logger.instance.class.to_s
+
+          compact_hash(config_options_to_keep)
+        end
+
+        def format_configuration_value(value)
+          # TODO: If the telemetry spec is updated to accept floats, this condition should be removed
+          if value.is_a?(Float) || value.is_a?(Array)
+            value.to_s
+          else
+            value
+          end
+        end
+
+        def compact_hash(hash)
+          hash.delete_if { |_k, v| v.nil? }
+        end
+
+        def env
+          Datadog.configuration.env
+        end
+
+        def service_name
+          Datadog.configuration.service
+        end
+
+        def service_version
+          Datadog.configuration.version
+        end
+
+        def tracer_version
+          Core::Environment::Identity.tracer_version
+        end
+
+        def products
+          profiler_obj = profiler
+          appsec_obj = appsec
+          profiler_obj || appsec_obj ? Telemetry::V1::Product.new(profiler: profiler_obj, appsec: appsec_obj) : nil
+        end
+
+        def profiler
+          version = profiler_version
+          Telemetry::V1::Profiler.new(version: version) if version
+        end
+
+        def profiler_version
+          tracer_version if Datadog.configuration.respond_to?(:profiling) && Datadog.configuration.profiling.enabled
+        end
+
+        def appsec
+          version = appsec_version
+          Telemetry::V1::AppSec.new(version: version) if version
+        end
+
+        def appsec_version
+          tracer_version if Datadog.configuration.respond_to?(:appsec) && Datadog.configuration.appsec.enabled
```

Comment:
No need to check `respond_to?` here. The configuration will always respond to `:appsec`. If it doesn't unit tests for this class will catch that.